Reintroduce Address type for package sets

colinwahl opened the pull request:
Intends to close #245 

This PR reintroduces the `Address` type, which is used for overriding a package in a package set. It allows users to replace individual packages within a package set, or add their own, without that package being changed in the upstream package set.

So far, the following is done:
- [x] Reintroduce `Address.dhall`, with improvements to the `Repo` constructor
- [x] Reintroduce documentation for overriding a package in a package set
- [x] Migrate example package set to use `Address`
- [x] Add a `PackageSet.dhall` type specification
- [x] Add an `assert` function to ensure official package sets only rely on packages within the registry

Notably the `assert` check to ensure our official package set is only made up of `Registry` entries has not been implemented. I understand how the check works/what it is for - but I am not sure the best place to put it. 

@f-f - two questions for you. First, is having the `PackageSet.dhall` specification useful? I noticed I couldn't type the example package set as that type, presumably because the keys of the map are unknown, but my dhall-fu isn't quite there yet. Second - where do you think the `assert` should live? I was thinking it could go in `PackageSet.dhall` and be imported to the package sets when we construct them. I suggest this because I don't think it'd be ideal to have to that implement in every package set.

thomashoneyman commented on v1/sets/20200911.dhall:
```diff
@@ -414,4 +412,8 @@ let packages =
 
+let isPackageSetInRegistry = ../isPackageSetInRegistry.dhall
+
+let test = assert : isPackageSetInRegistry (toMap packages) === True
```
Should this test be included in every package set we publish?

f-f replied:
Yes. The `assert` operator can't be packaged inside a function (because you can't do assertions about abstract values, only concrete ones), so this line goes in every set. Everything else is self-contained in that function we import

thomashoneyman commented on v1/isPackageSetInRegistry.dhall:
```diff
@@ -0,0 +1,23 @@
+      Prelude.List.all Bool (\(x : Bool) -> x == True)
+        (Prelude.List.map Pkg Bool isRegistryPkg pkgs)
+
+in isPackageSetInRegistry
```
```suggestion
in isPackageSetInRegistry

```

thomashoneyman commented on README.md:
```diff
@@ -228,6 +228,39 @@ but it's allowed only for a set period of time because of the `leftpad` problem
+
+let overrides = { effect = From.Local ( ../my-effect/spago.dhall as Location ) }
+
+in  Registry.Package::{
+    , name = "my-package"
+    , packages = 
+        Registry.Index.PackageSet
+          { compiler = upstream.compiler
+          , packages = toMap (upstream.packages // overrides)
+          }
+    , targets =
+        toMap
+          { src = Registry.Target::{
+            , sources = [ "src/**/*.purs" ]
+            , dependencies = toMap { effect = "^2.0.0" }
+            }
+          }
+    }
```
@f-f I wonder if there's an easier way to specify an override, as this involves a bit of specific knowledge (`toMap`, the type annotations) that I'd rather not force users to learn about. Is there a way to abstract any of this out?

f-f replied:
A new Dhall operator has been proposed to deal with these cases (changing a record without using `toMap`): https://github.com/dhall-lang/dhall-lang/issues/979
The status there is that there's a proposal, there's pretty much agreement that it's good to have, there's availability to put a bounty on it, and it needs someone to implement it.

In any case, I would not add this documentation over here, since it doesn't really concern the registry?

colinwahl replied:
Oh hah, I had meant this to be a comment on how to override a package within a `PackageSet`, but misread the old documentation when I moved it over. I'll change this to be an example on how to override a specific package within a package set

f-f replied:
Yeah, I think we only need documentation about how to author such package sets

f-f commented:
@colinwahl thanks for doing this!

A couple of notes:
- I just pushed a commit with the implementation of the `assert` check - we can abstract away most of it, but we'll have to copy one line in every package set we have
- as you realized it's not possible to give a proper type to the package set because it's a record. We might want to make sure that the set has the right type - in that case we could run `toMap` (as we do for the assertion) and then conform to the `PackageSet` type that you added (we'll just need a type annotation) if we don't need the `PackageSet`

f-f commented:
Oh, I forgot to bring this up: IIRC the reason we have the form
```dhall
prelude = Address.Registry { name = "prelude", version = "v6.0.0" }
```
instead of 
```dhall
prelude = Address.Registry { version = "v6.0.0" }
```
..is so that the pointer to the package is "self-contained", i.e. we have all the information to get to the package in the record, instead of having to rely on the key.

Now, the question I have is: is this useful? Self-contained is nice, but the second form is much nicer to look at and to deal with? (less accidental typos, etc.) Should we switch to the second form? cc @thomashoneyman

thomashoneyman commented:
Since users also use the `Address` type in their code, I would greatly prefer the second form:

```dhall
prelude = Address.Registry { version = "v6.0.0" }
```

or even

```dhall
prelude = Address.Registry "v6.0.0"
```

thomashoneyman commented:
I'm struggling to think of a time when we would need the self-contained version and wouldn't be able to produce it from the not-self-contained version. In PureScript we can easily do this processing step, and I'd assume that even in Dhall it should be possible to turn a `Map Package Version` into a `Map Package { package, version }`

colinwahl commented:
Got it - I'll make the change to the `Address.Registry` constructor.

colinwahl commented:
Alrighty - I've updated the `Address.Registry` construct to not include the `name`, and I've corrected the documentation.

I looked a bit into adding the `PackageSet` type back in - but I decided it wasn't worth it at this point. We can loop back and change that later if we want!

f-f commented on v1/Address.dhall:
```diff
@@ -0,0 +1,19 @@
+let Repo = ./Repo.dhall
+
+let Address =
+    < Registry : { version : Text }
```
Should we change this to
```suggestion
    < Registry : Text
```
as @thomashoneyman suggested?

colinwahl replied:
As discussed in the Registry call - we will introduce a `SemVer = Text` type alias and use that directly in the `Registry` constructor.

thomashoneyman replied:
Once that's done this looks good to me

thomashoneyman commented on v1/isPackageSetInRegistry.dhall:
```diff
@@ -9,7 +9,7 @@ let Repo = ./Repo.dhall
 let Pkgs = Prelude.Map.Type Text Address
 let isRegistryPkg = \(pkg : Pkg) -> merge
-      { Registry = \(x : { version : Text }) -> True
+      { Registry = \(x : Text) -> True
```
Should this be SemVer?

f-f replied:
The SemVer synonym is not defined in this file, so we'd have to define it again or pull it out as a separate definition, but I don't think any of these are worth doing (as one can always have a look at where `Address` is defined)

f-f reviewed (approved):
Thanks @colinwahl!
